src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,328 @@
+/**
+ * @file gan.hpp
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_IMPL_HPP
+#define MLPACK_METHODS_ANN_GAN_IMPL_HPP
+
+#include "gan.hpp"
+
+#include <mlpack/core.hpp>
+#include <chrono>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+#include <assert.h>
+
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/core/dists/gaussian_distribution.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::math;
+using namespace mlpack::distribution;
+
+namespace mlpack {
+namespace ann /** artifical neural network  */ {
+template<typename Generator, typename Discriminator, typename IntializerType>
+GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::GenerativeAdversarialNetwork(arma::mat& trainData,
+    IntializerType initializeRule,
+    Generator& generator,
+    Discriminator& discriminator,
+    size_t batchSize,
+    size_t iterations,
+    size_t disIteration,
+    size_t generatorInSize):
+    trainData(trainData),
+    initializeRule(initializeRule),
+    generator(generator),
+    discriminator(discriminator),
+    trainGenerator(false),
+    batchSize(batchSize),
+    iterations(iterations),
+    disIteration(disIteration),
+    generatorInSize(generatorInSize),
+    reset(false)
+{
+  discriminator.network.insert(discriminator.network.begin(), new Join<>());
```

Comment:
If I am not mistaken, you add this line in order get the error, right? In such a way I suggest `IdentityLayer` (see 'ann/layer/base_layer.hpp').